pkg/controllers/nodeclass/termination/controller.go

```diff
@@ -0,0 +1,149 @@
+/*
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package termination
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"k8s.io/apimachinery/pkg/api/errors"
+
+	nodeclassevents "github.com/aws/karpenter-provider-aws/pkg/controllers/nodeclass/events"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/launchtemplate"
+
+	"github.com/samber/lo"
+	"golang.org/x/time/rate"
+	"k8s.io/apimachinery/pkg/api/equality"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/util/workqueue"
+	controllerruntime "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/manager"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	corev1beta1 "sigs.k8s.io/karpenter/pkg/apis/v1beta1"
+	"sigs.k8s.io/karpenter/pkg/events"
+	corecontroller "sigs.k8s.io/karpenter/pkg/operator/controller"
+
+	"github.com/aws/karpenter-provider-aws/pkg/apis/v1beta1"
+	"github.com/aws/karpenter-provider-aws/pkg/providers/instanceprofile"
+)
+
+var _ corecontroller.FinalizingTypedController[*v1beta1.EC2NodeClass] = (*Controller)(nil)
+
+type Controller struct {
+	kubeClient              client.Client
+	recorder                events.Recorder
+	instanceProfileProvider instanceprofile.Provider
+	launchTemplateProvider  launchtemplate.Provider
+}
+
+func NewController(kubeClient client.Client, recorder events.Recorder,
+	instanceProfileProvider instanceprofile.Provider, launchTemplateProvider launchtemplate.Provider) corecontroller.Controller {
+
+	return corecontroller.Typed[*v1beta1.EC2NodeClass](kubeClient, &Controller{
+		kubeClient:              kubeClient,
+		recorder:                recorder,
+		instanceProfileProvider: instanceProfileProvider,
+		launchTemplateProvider:  launchTemplateProvider,
+	})
+}
+
+func (c *Controller) Reconcile(ctx context.Context, nodeClass *v1beta1.EC2NodeClass) (reconcile.Result, error) {
+	stored := nodeClass.DeepCopy()
+	if controllerutil.ContainsFinalizer(nodeClass, v1beta1.TerminationFinalizer) {
```

Comment:
This is going to cause leaks. The finalizer needs to be directly tied to the operation that it's blocking in its termination flow. In the way that you have it here, it wouldn't happen often, but it's possible that the finalizer never gets patched on, there was a NodeClaim launch, and then the EC2NodeClass is terminated. This will never wait for the launch templates to cleanup or the instance profiles to go away.

Reply:
Discussed offline, we will be adding the NodeClass finalizer as part of the instanceProfile Creation. Since the instanceProfile creation is current operation that gates instance launch, that will prevent the race of when EC2NodeClass does not have a finalizer and gets deleted.